src/apis/server.rs

```diff
@@ -1,15 +1,156 @@
-use std::net::SocketAddr;
+use std::net::{SocketAddr, TcpListener};
 use std::sync::Arc;
 
 use axum_server::tls_rustls::RustlsConfig;
 use axum_server::Handle;
 use futures::Future;
 use log::info;
+use tokio::task::JoinHandle;
 use warp::hyper;
 
 use super::routes::router;
+use crate::signals::shutdown_signal_with_message;
 use crate::tracker::Tracker;
 
+#[derive(Debug)]
+pub enum Error {
+    Error(String),
+}
+
+#[allow(clippy::module_name_repetitions)]
+pub type StoppedApiServer = ApiServer<Stopped>;
+#[allow(clippy::module_name_repetitions)]
+pub type RunningApiServer = ApiServer<Running>;
+
+#[allow(clippy::module_name_repetitions)]
+pub struct ApiServer<S> {
+    pub cfg: torrust_tracker_configuration::HttpApi,
+    pub tracker: Arc<Tracker>,
+    pub state: S,
+}
+
+pub struct Stopped;
+
+pub struct Running {
+    pub bind_address: SocketAddr,
+    stop_job_sender: tokio::sync::oneshot::Sender<u8>,
+    job: JoinHandle<()>,
+}
+
+impl ApiServer<Stopped> {
+    pub fn new(cfg: torrust_tracker_configuration::HttpApi, tracker: Arc<Tracker>) -> Self {
+        Self {
+            cfg,
+            tracker,
+            state: Stopped {},
+        }
+    }
+
+    /// # Errors
+    ///
+    /// Will return `Err` if `TcpListener` can not bind to `bind_address`.
+    pub fn start(self) -> Result<ApiServer<Running>, Error> {
+        let listener = TcpListener::bind(&self.cfg.bind_address).map_err(|e| Error::Error(e.to_string()))?;
+
+        let bind_address = listener.local_addr().map_err(|e| Error::Error(e.to_string()))?;
+
+        let cfg = self.cfg.clone();
+        let tracker = self.tracker.clone();
+
+        let (sender, receiver) = tokio::sync::oneshot::channel::<u8>();
+
+        let job = tokio::spawn(async move {
+            if let (true, Some(ssl_cert_path), Some(ssl_key_path)) = (cfg.ssl_enabled, cfg.ssl_cert_path, cfg.ssl_key_path) {
+                let tls_config = RustlsConfig::from_pem_file(ssl_cert_path, ssl_key_path)
+                    .await
+                    .expect("Could not read ssl cert and/or key.");
+
+                start_tls_from_tcp_listener_with_graceful_shutdown(listener, tls_config, &tracker, receiver)
+                    .await
+                    .expect("Could not start from tcp listener with tls.");
+            } else {
+                start_from_tcp_listener_with_graceful_shutdown(listener, &tracker, receiver)
+                    .await
+                    .expect("Could not start from tcp listener.");
+            }
+        });
```

Comment:
Hi @josecelano ,

Your question has kept me occupied for an hour xd. So the problem with these tokio spawns is that they will not abort the parent program when a panic occurs. So we won't actually know if there was a panic in a task.

You solved this partially [here](https://github.com/torrust/torrust-tracker/blob/develop/src/jobs/tracker_apis.rs#L50) by using a channel to report whether the api server has started. The problem with our current code however is that it will only check whether the `RusttlsConfig` was created successfully and not if the actual server has started successfully.

We could add another channel that will report something when the await on the server start function has yielded an error. This way we can wait a little bit after starting the server start job/tokio task to see whether an error has been sent through this channel.

So to answer your questions:
> Has the behaviour changed when you cannot start the server with TSL?
Yes and I need to fix it :). 

> Does the expect means the whole application is going to halt when the TLS configuration is missing? If that's true, should not we just warn the user and try to start other services?
Since the `expect` will yield a panic on a tokio task, only the task will abort, but the rest of the app will continue like nothing happened.

Reply:
Hi @WarmBeer, thank you for the clarification!

I think we should start all or nothing too. People don't read logs :-). If something fails and you don't care, you should explicitly disable it.

In summary, we need to change even the previous behaviour if it only shows a warning.

This is getting more and more complex. I do not know if we should add a second one-shot channel or maybe a normal channel and define a list of messages between the launcher and the "launched" thing. Alternatively, we could try to check if the service is alive and ready just using the public contract. For example, for the API, we could check if the `status` endpoint is returning a `200`, but I do not like this solution because the launcher uses a "polling" approach. I think a proper communication channel between the laucher and the child is the right solution.

Reply:
@josecelano 

I will restore the current "develop" behaviour in this PR and will open a new (issue)[https://github.com/torrust/torrust-tracker/issues/213] for how we are going to deal with this.

Reply:
OK @WarmBeer. I think we could replace the warning with an error, at least.